src/common/stats/StatsManager.cpp

```diff
@@ -309,6 +309,245 @@ StatusOr<StatsManager::VT> StatsManager::readHisto(const std::string& counterNam
+void PrometheusSerializer::WriteValue(std::ostream& out, double value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        auto saved_flags = out.setf(std::ios::fixed, std::ios::floatfield);
+        out << value;
+        out.setf(saved_flags, std::ios::floatfield);
+    }
+}
+
+void PrometheusSerializer::WriteValue(std::ostream& out, const int64_t value) const {
+    if (std::isnan(value)) {
+        out << "Nan";
+    } else if (std::isinf(value)) {
+        out << (value < 0 ? "-Inf" : "+Inf");
+    } else {
+        out << value;
+    }
+}
+
+void PrometheusSerializer::WriteValue(std::ostream& out, const std::string& value) const {
+    for (auto c : value) {
+        switch (c) {
+        case '\n':
+            out << '\\' << 'n';
+            break;
+
+        case '\\':
+            out << '\\' << c;
+            break;
+
+        case '"':
+            out << '\\' << c;
+            break;
+
+        default:
+            out << c;
+            break;
+        }
+    }
+}
+
+template<typename T>
+void PrometheusSerializer::WriteHead(std::ostream& out, const std::string& family,
+                const std::vector<Label>& labels, const std::string& suffix,
+                const std::string& extraLabelName,
+                const T extraLabelValue) const {
+    out << family << suffix;
+    if (!labels.empty() || !extraLabelName.empty()) {
+        out << "{";
+        const char* prefix = "";
+
+        for (auto& lp : labels) {
+            out << prefix << lp.name << "=\"";
+            WriteValue(out, lp.value);
+            out << "\"";
+            prefix = ",";
+        }
+        if (!extraLabelName.empty()) {
+            out << prefix << extraLabelName << "=\"";
+            WriteValue(out, extraLabelValue);
+            out << "\"";
+        }
+        out << "}";
+    }
+    out << " ";
+}
+
+// Write a line trailer: timestamp
+void PrometheusSerializer::WriteTail(std::ostream& out, const std::int64_t timestamp_ms) const {
+    if (timestamp_ms != 0) {
+        out << " " << timestamp_ms;
+    }
+    out << "\n";
+}
+
+// TODO(shylock) transform nebula stats to Prometheus::ClientMetric/MetricFamily
+// using the serrializing from 3rd-SDK
+void StatsManager::PrometheusSerialize(std::ostream& out) /*const*/ {
+    std::locale saved_locale = out.getloc();
+    out.imbue(std::locale::classic());
+    for (auto& index : nameMap_) {
+        auto parsedName = parseMetricName(index.first);
+        std::string name = parsedName.ok() ? parsedName.value().name : index.first;
+        if (isStatIndex(index.second)) {
+            Annotate(out, name, "gauge");
+
+            WriteHead(out, name, {});
+            if (parsedName.ok()) {
+                // make sure contains value
+                WriteValue(out,
+                    readStats(index.second, TimeRange::ONE_MINUTE, StatsMethod::AVG).value());
```

Comment:
Why avg in this line and sum in line 491?

Reply:
Thanks.It's a mistake.